### PR TITLE
2560: fail-fast when trace vars mismatches model vars

### PR DESCRIPTION
**Pull request by jazzmuesli**
I'm not sure about testing, I reproduced mine in 2560.py - but I guess it's better to add it in pymc3/tests//test_text_backend.py. Do I really want to create text files in unit tests though?

### CODE REVIEW

**Comment by springcoil** [comment]
#2560 This largely looks good to me. Some suggestions are to create a regression test with your .py file in it.

File: 2560.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import pymc3 as pm
+with pm.Model() as model:
```
Comment (ColCarroll) [inline 2560.py:2]: this should be a function in `pymc3/tests/test_text_backend.py`.  Something like

```
def test_wont_dump_mismatched_vars():
    with pm.Model() as model:
        a = pm.Gamma('a', mu=10.0, sd=2.0)
        b = pm.Gamma('b', mu=10.0, sd=2.0)

        partial_trace = pm.sample(draws=100, tune=0, trace=[model.a, model.a_log__])
        assert len(partial_trace.varnames) == 2
    
        with pytest.raises(ValueError):
            pm.backends.text.dump('trace.text', partial_trace)
```

File: pymc3/backends/text.py [hunk 2]
```diff
@@ -200,6 +201,12 @@ def dump(name, trace, chains=None):
         os.mkdir(name)
     if chains is None:
         chains = trace.chains
+    model = modelcontext(model)
+    model_vars = sorted([x.name for x in model.unobserved_RVs])
+    trace_vars = sorted(trace.varnames)
+
+    if model_vars != trace_vars:
+        raise ValueError('Variables mismatch: model_vars=' + str(model_vars) + ", trace_vars=" + str(trace_vars))
```
Comment (ColCarroll) [inline pymc3/backends/text.py:209]: This error might be more explicit:
```
ValueError('Cannot serialize a trace that does not include all model variables.'
           '\n\tmodel variables: {}'
           '\n\ttrace variables: {}'.format(', '.join(model_vars), ', '.join(trace_vars))
```

**Review by ColCarroll — COMMENTED** [review]
Thanks for this -- it is a funny problem, and I think this is a helpful fix.  I left perhaps overly specific feedback on the test, since it can be hard to navigate our test suite!

**Comment by springcoil** [comment]
Valid points @colcarroll. I too find the tests hard to navigate.

**Comment by fonnesbeck** [comment]
Does this PR have legs, or can it be closed?

**Comment by springcoil** [comment]
I think it can be closed

**Pull request closed without merging**